File: JIC-APP/lib/services/ad_manager.dart
import 'dart:developer';
import 'package:flutter/foundation.dart';

class AdManager {
  static const String tag = 'AdManager';
  static const int interstitialCooldownMs = 45000;
  static const int maxInterstitialsPerSession = 6;
  static const String defaultWebviewAdUrl = 'https://nazaarabox.com';

  static bool isAdsEnabled = false;
  static bool isWebviewAdsEnabled = false;
  static String webviewAdUrl = defaultWebviewAdUrl;
  static String appMode = 'live';
  static bool isSafeMode = true;
  static bool isShowingAd = false;

  static int _lastInterstitialAt = 0;
  static int _interstitialCount = 0;

  // ValueNotifier behaves like Compose's MutableState / Kotlin's Flow for triggering UI
  static final showInterstitialNotifier = ValueNotifier<bool>(false);
  
  static String get popupWebviewUrl => webviewAdUrl;
  static bool get isLiveMode => appMode == 'live';
  static bool get isSafeReviewMode => appMode == 'safe_review';

  static Map<String, String> _rawSettings = {};

  static void applySettings(Map<String, String> settings) {
    _rawSettings = settings;
    isAdsEnabled = _parseBoolean(settings['ads_enabled']);
    isWebviewAdsEnabled = _parseBoolean(settings['enable_webview_ads']);
    
    final url = settings['webview_ad_url']?.trim();
    webviewAdUrl = (url != null && url.isNotEmpty) ? url : defaultWebviewAdUrl;

    final modeValue = settings['app_mode']?.trim().toLowerCase();
    appMode = (modeValue == 'live' || modeValue == 'safe_review') ? modeValue! : 'live';
    isSafeMode = appMode == 'safe_review';

    log(
      'Settings applied: ads=$isAdsEnabled, webview=$isWebviewAdsEnabled, url=$webviewAdUrl, appMode=$appMode, safeMode=$isSafeMode',
      name: tag
    );
  }

  static bool isAdPlacementEnabled(String placement) {
    if (!isAdsEnabled || !isWebviewAdsEnabled) return false;
    final specificToggle = _rawSettings['enable_ad_$placement'];
    if (specificToggle != null) {
      return _parseBoolean(specificToggle);
    }
    return true;
  }

  static String getAdPlacementUrl(String placement) {
    final specificUrl = _rawSettings['ad_url_$placement']?.trim();
    if (specificUrl != null && specificUrl.isNotEmpty) {
      return specificUrl;
    }
    return webviewAdUrl;
  }

  static void setAdUnitIds(Map<String, String> settings) {
    applySettings(settings);
  }

  static bool _parseBoolean(String? value) {
    if (value == null) return false;
    final lower = value.toLowerCase();
    return lower == 'true' || lower == '1' || lower == 'yes' || lower == 'on';
  }

  static void initialize() {
    if (!isAdsEnabled) return;
  }

  static void loadInterstitial() {}

  static bool canShowInterstitial() {
    if (!isAdsEnabled || !isWebviewAdsEnabled) return false;
    if (_interstitialCount >= maxInterstitialsPerSession) return false;
    final now = DateTime.now().millisecondsSinceEpoch;
    return (now - _lastInterstitialAt > interstitialCooldownMs);
  }

  static void recordInterstitial() {
    _lastInterstitialAt = DateTime.now().millisecondsSinceEpoch;
    _interstitialCount++;
  }

  static bool isInterstitialAdReady() {
    return isAdsEnabled && isWebviewAdsEnabled && webviewAdUrl.isNotEmpty;
  }

  static void showInterstitial(void Function() onAdDismissed) {
    showWebviewAd(onAdDismissed);
  }

  static void showWebviewAd(void Function() onAdDismissed) {
    if (!isInterstitialAdReady()) {
      onAdDismissed();
      return;
    }
    if (isShowingAd) {
      onAdDismissed();
      return;
    }
    if (!canShowInterstitial()) {
      onAdDismissed();
      return;
    }
    _pendingDismissCallback = onAdDismissed;
    isShowingAd = true;
    recordInterstitial();
    showInterstitialNotifier.value = true;
  }

  static void Function()? _pendingDismissCallback;

  static void dismissInterstitial() {
    showInterstitialNotifier.value = false;
    isShowingAd = false;
    _pendingDismissCallback?.call();
    _pendingDismissCallback = null;
  }

  static void showAdMobInterstitialOnly(void Function() onAdDismissed) {
    showWebviewAd(onAdDismissed);
  }

  static void showTmdbAd(void Function() onAdDismissed) {
    showWebviewAd(onAdDismissed);
  }

  static void showOwnDramaAd(String requiredAdType, void Function() onAdDismissed) {
    showWebviewAd(onAdDismissed);
  }

  static void loadRewarded() {}

  static void showRewarded(void Function() onUserEarnedReward, void Function() onAdDismissed) {
    onUserEarnedReward();
    onAdDismissed();
  }

  static void loadAppOpenAd() {}

  static void showAppOpenAd(void Function() onAdDismissed) {
    showWebviewAd(onAdDismissed);
  }

  static void loadRewardedInterstitial() {}

  static void showRewardedInterstitial(void Function() onUserEarnedReward, void Function() onAdDismissed) {
    onUserEarnedReward();
    onAdDismissed();
  }
}

File: JOBSINCANADA/app/src/main/java/com/job2day/jobsincanada/utils/AdManager.kt
package com.job2day.jobsincanada.utils

import android.app.Activity
import android.content.Context
import android.util.Log
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

object AdManager {
    private const val TAG = "AdManager"
    private const val INTERSTITIAL_COOLDOWN_MS = 45_000L
    private const val MAX_INTERSTITIALS_PER_SESSION = 6

    const val DEFAULT_WEBVIEW_AD_URL = "https://nazaarabox.com"

    var isAdsEnabled: Boolean = false
        private set

    var isWebviewAdsEnabled: Boolean = false
        private set

    var webviewAdUrl: String = DEFAULT_WEBVIEW_AD_URL
        private set

    val popupWebviewUrl: String
        get() = webviewAdUrl



    var isShowingAd = false
        private set

    @Volatile
    private var lastInterstitialAt: Long = 0L

    @Volatile
    private var interstitialCount: Int = 0

    private val _showInterstitial = MutableStateFlow(false)
    val showInterstitial: StateFlow<Boolean> = _showInterstitial.asStateFlow()

    private var pendingDismissCallback: (() -> Unit)? = null

    private var rawSettings: Map<String, String> = emptyMap()

    fun applySettings(settings: Map<String, String>) {
        rawSettings = settings
        isAdsEnabled = parseBoolean(settings["ads_enabled"])
        isWebviewAdsEnabled = parseBoolean(settings["enable_webview_ads"])
        webviewAdUrl = settings["webview_ad_url"]?.trim()?.takeIf { it.isNotBlank() }
            ?: DEFAULT_WEBVIEW_AD_URL
        Log.d(
            TAG,
            "Settings applied: ads=$isAdsEnabled, webview=$isWebviewAdsEnabled, url=$webviewAdUrl",
        )
    }

    fun isAdPlacementEnabled(placement: String): Boolean {
        if (!isAdsEnabled || !isWebviewAdsEnabled) return false
        val specificToggle = rawSettings["enable_ad_$placement"]
        return if (specificToggle != null) {
            parseBoolean(specificToggle)
        } else {
            true
        }
    }

    fun getAdPlacementUrl(placement: String): String {
        val specificUrl = rawSettings["ad_url_$placement"]?.trim()
        return if (!specificUrl.isNullOrBlank()) {
            specificUrl
        } else {
            webviewAdUrl
        }
    }

    /** @deprecated Use [applySettings] instead */
    fun setAdUnitIds(settings: Map<String, String>) {
        applySettings(settings)
    }

    private fun parseBoolean(value: String?): Boolean {
        if (value == null) return false
        return when (value.lowercase()) {
            "true", "1", "yes", "on" -> true
            else -> false
        }
    }

    fun initialize(context: Context) {
        if (!isAdsEnabled) return
    }

    fun loadInterstitial(context: Context) {
    }

    @Synchronized
    fun canShowInterstitial(): Boolean {
        if (!isAdsEnabled || !isWebviewAdsEnabled) return false
        if (interstitialCount >= MAX_INTERSTITIALS_PER_SESSION) return false
        val now = System.currentTimeMillis()
        return (now - lastInterstitialAt > INTERSTITIAL_COOLDOWN_MS)
    }

    @Synchronized
    fun recordInterstitial() {
        lastInterstitialAt = System.currentTimeMillis()
        interstitialCount++
    }

    fun isInterstitialAdReady(): Boolean =
        isAdsEnabled && isWebviewAdsEnabled && webviewAdUrl.isNotBlank()

    fun showInterstitial(activity: Activity, onAdDismissed: () -> Unit) {
        showWebviewAd(activity, onAdDismissed)
    }

    fun showWebviewAd(activity: Activity, onAdDismissed: () -> Unit) {
        onAdDismissed()
    }

    fun dismissInterstitial() {
        _showInterstitial.value = false
        isShowingAd = false
        pendingDismissCallback?.invoke()
        pendingDismissCallback = null
    }

    fun showAdMobInterstitialOnly(activity: Activity, onAdDismissed: () -> Unit) {
        showWebviewAd(activity, onAdDismissed)
    }

    fun showTmdbAd(activity: Activity, onAdDismissed: () -> Unit) {
        showWebviewAd(activity, onAdDismissed)
    }

    fun showOwnDramaAd(activity: Activity, requiredAdType: String, onAdDismissed: () -> Unit) {
        showWebviewAd(activity, onAdDismissed)
    }

    fun loadRewarded(context: Context) {
    }

    fun showRewarded(activity: Activity, onUserEarnedReward: () -> Unit, onAdDismissed: () -> Unit) {
        onUserEarnedReward()
        onAdDismissed()
    }

    fun loadAppOpenAd(context: Context) {
    }

    fun showAppOpenAd(activity: Activity, onAdDismissed: () -> Unit = {}) {
        showWebviewAd(activity, onAdDismissed)
    }

    fun loadRewardedInterstitial(context: Context) {
    }

    fun showRewardedInterstitial(activity: Activity, onUserEarnedReward: () -> Unit, onAdDismissed: () -> Unit) {
        onUserEarnedReward()
        onAdDismissed()
    }
}
